src/Migrator/Migrator.php

```diff
@@ -0,0 +1,267 @@
+        $this->destConnection = $destConnection;
+        $this->logger = $logger;
+        $this->referenceStore = new ReferenceStore();
+    }
+
+    /**
+     * Perform the configured migrations
+     *
+     * @throws \Exception Rethrows any exceptions after logging
+     */
+    public function execute(string $setName, TableCollection $tableCollection, ViewCollection $viewCollection): void
+    {
+        foreach ($tableCollection->getTables() as $table => $migrationSpec) {
+            // @todo: it'd probably be better to have a proper `migrationspec` config object
+            // rather than relying on properties being present in the json / stdClass object
+
+            $sampler = $this->buildTableSampler($migrationSpec, $table);
+            $writer = new Writer($migrationSpec, $this->destConnection);
+            $cleaner = new RowCleaner($migrationSpec);
+
+            foreach ($this->customCleaners as $alias => $customCleaner) {
+                $cleaner->registerCleaner($customCleaner, $alias);
+            }
+
+            try {
+                $this->ensureEmptyTargetTable($table, $this->sourceConnection, $this->destConnection);
+                $rows = $sampler->execute();
+
+                foreach ($rows as $row) {
+                    $writer->write($table, $cleaner->cleanRow($row));
+                }
+                $writer->postWrite();
+
+                $this->logger->info("$setName: migrated '$table' with '" . $sampler->getName() . "': " . \count($rows) . " rows");
+            } catch (\Exception $e) {
+                $this->logger->error(
+                    "$setName: failed to migrate '$table' with '" . $sampler->getName() . "': " . $e->getMessage()
+                );
+                throw $e;
+            }
+        }
+
+        foreach ($viewCollection->getViews() as $view) {
+            $this->migrateView($view, $setName);
+        }
+
+        $this->migrateTableTriggers($setName, $tableCollection);
+    }
+
+    public function registerCustomCleaner(FieldCleaner $cleaner, string $alias): void
+    {
+        $this->customCleaners[$alias] = $cleaner;
+    }
+
+    /**
+     * Ensure that the specified table is present in the destination DB as an empty copy of the source
+     *
+     * @param string $table Table name
+     * @param Connection $sourceConnection Originating DB connection
+     * @param Connection $destConnection Target DB connection
+     *
+     * @return void
+     * @throws \RuntimeException If DB type not supported
+     * @throws \Doctrine\DBAL\DBALException If target table cannot be removed or recreated
+     */
+    private function ensureEmptyTargetTable($table, Connection $sourceConnection, Connection $destConnection)
+    {
+        // SchemaManager doesn't do enums!
+        $destConnection->exec('DROP TABLE IF EXISTS ' . $sourceConnection->quoteIdentifier($table));
+
+        $driverName = $sourceConnection->getDriver()->getName();
+        if ($driverName === 'pdo_mysql') {
```

Comment:
in a future PR i think i'd like to add a `DatabaseDriver` interface, with separate concrete classes for `MySQL` and `SQLite`. This would make supporting other db drivers (such as Postgres) far easier